aten/src/ATen/native/quantized/cudnn/Conv.cpp

```diff
@@ -0,0 +1,360 @@
+}
+
+//
+// output Tensor will be a fp32 Tensor
+// both act and weight will be int8 Tensor
+//
+template <int kSpatialDim>
+Tensor raw_cudnn_convolution_forward(
+    const Tensor& act,
+    const Tensor& weight,
+    IntArrayRef padding,
+    IntArrayRef stride,
+    IntArrayRef dilation,
+    int64_t groups,
+    bool benchmark,
+    bool deterministic,
+    bool allow_tf32,
+    float requantize_multiplier) {
+  // TODO: add dimension validations for input/weight/bias
+  const int N = act.size(0);
+  const int C = act.size(1);
+  const int D = kSpatialDim == 3 ? act.size(2) : 1;
+  const int H = act.size(kSpatialDim);
+  const int W = act.size(kSpatialDim + 1);
+  const int M = weight.size(0); // output channels
+  std::vector<int64_t> kernel_size = {weight.size(2), weight.size(3)};
+  at::SmallVector<int64_t, kSpatialDim + 2> output_shape;
+  output_shape = MakeConvOutputShape<kSpatialDim>(N, M, {H, W}, kernel_size, stride, padding, dilation);
+  Tensor output_fp32 = at::empty(
+      output_shape,
+      at::device(at::kCUDA).dtype(at::kFloat),
+      at::MemoryFormat::ChannelsLast
+  );
+
+  raw_cudnn_convolution_forward_out(
+      output_fp32, act, weight,
+      padding, stride, dilation, groups,
+      benchmark,
+      deterministic,
+      allow_tf32,
+      requantize_multiplier);
+  return output_fp32;
+}
+
+
+template <int kSpatialDim, bool kReluFused>
+class QConvInt8 final {
+ public:
+  static Tensor run(
+      Tensor act,
+      Tensor weight,
+      c10::optional<Tensor> bias,
+      torch::List<int64_t> stride,
+      torch::List<int64_t> padding,
+      torch::List<int64_t> dilation,
+      int64_t groups,
+      double output_scale,
+      int64_t output_zero_point) {
+    TORCH_CHECK(!kReluFused, "conv relu not supported yet");
+    TORCH_CHECK(!bias.has_value(), "bias is not supported yet");
+    act = act.contiguous(c10::MemoryFormat::ChannelsLast);
+    weight = weight.contiguous(c10::MemoryFormat::ChannelsLast);
+    // requantization
+    // out_int8 = act_int8 * weight_int8 * act_scale * w_scale / output_scale
+    auto act_scale = act.q_scale();
+    auto weight_scale = weight.q_scale();
+    auto requantize_multiplier = act_scale * weight_scale / output_scale;
+    //auto requantize_multiplier = output_scale / (act_scale * weight_scale);
+    // TODO: check all zero_points are zero/all tensors are symmetrically quantized
+    Tensor output_fp32_requant = raw_cudnn_convolution_forward<kSpatialDim>(
+        act.int_repr(), weight.int_repr(),
+        IntArrayRef(padding.vec()), IntArrayRef(stride.vec()), IntArrayRef(dilation.vec()), groups,
+        false /* benchmark */,
+        true /* deterministic */,
+        false /* allow_tf32 */,
+        requantize_multiplier
+    );
+
+    // convert output fp32 Tensor to int8 by clamping
+    // TODO: get the range based on target output dtype, for now
+    // we hardcode this to the range for int8
+    Tensor clampped = output_fp32_requant.clamp(-128, 127);
```

Comment:
not sure I understand this line, how does "convert fp32 to int8 by clamping" work?

Reply:
oh we first clamp, the type conversion is done in L347 in `clampped.to(at::kChar)`, maybe we can have a separate Tensor to make it clearer